### ISSUE TITLE
curl_easy_perform blocking infinitely

### ISSUE THREAD

**Issue by Jacellen** [curl/curl#12483]
### I did this

~~~c
size_t writeCallback(char* contents, size_t size, size_t nmemb, std::string* output) {
    size_t totalSize = size * nmemb;
    output->append(contents, totalSize);
    return totalSize;
}

void launcher::testing()
{
    curl_global_init(CURL_GLOBAL_DEFAULT);
    CURL* curl = curl_easy_init();
    if (curl) {
        std::string output;

        curl_easy_setopt(curl, CURLOPT_URL, "https://www.google.com/");
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &output);
        CURLcode res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

using RtlGetFullPathNameU_t = NTSTATUS(WINAPI*)(PCWSTR, ULONG, PWSTR, PULONG);
RtlGetFullPathNameU_t originalRtlGetFullPathNameU = nullptr;

ULONG WINAPI hookedRtlGetFullPathNameU(PCWSTR FileName, ULONG Size, PWSTR Buffer, PULONG ShortName)
{
	auto result = originalRtlGetFullPathNameU(FileName, Size, Buffer, ShortName);
	launcher::testing();
	return result;
}

bool hooks::initialized()
{
	if (MH_Initialize() != MH_OK) return false;
	if (MH_CreateHookApi(L"ntdll.dll", "RtlGetFullPathName_U", hookedRtlGetFullPathNameU, (void**)&originalRtlGetFullPathNameU) != MH_OK)return false;
	if (MH_EnableHook(nullptr) != MH_OK)return false;
	return true;
}
~~~

When executing this code it runs up to the curl_easy_perform and stays in there forever, never returning any value. When debugging, I found out that if I access the IP address directly, the request will pass, but in this code curl_easy_perform blocking infinitely.

### curl/libcurl version

last version

### operating system

win 11 23H2

**Comment by bagder** [comment]
You did not say which libcurl version you're using, and not which TLS library/version, but libcurl has used widely by many users on Windows. This is not likely to be a curl bug without more specific conditions applying. You need to help us find those.

**Comment by bagder** [comment]
> Nothing has changed

Then you did not recompile. Surely you got verbose output?

Did it connect? Then the connect timeout won't change anything.

>  curl_easy_setopt(curl, CURLOPT_ACCEPTTIMEOUT_MS, 5000);

That's not interesting for HTTP transfers. `CURLOPT_TIMEOUT` might be though.

**Comment by Jacellen** [comment]
> You did not say which libcurl version you're using, and not which TLS library/version, but libcurl has used widely by many users on Windows. This is not likely to be a curl bug without more specific conditions applying. You need to help us find those.

![image](https://github.com/curl/curl/assets/78690727/d48b792d-b04a-4f3f-816e-77c711471642)
8.5.0 version

**Comment by Jacellen** [comment]
> > Nothing has changed
> 
> Then you did not recompile. Surely you got verbose output?
> 
> Did it connect? Then the connect timeout won't change anything.
> 
> > curl_easy_setopt(curl, CURLOPT_ACCEPTTIMEOUT_MS, 5000);
> 
> That's not interesting for HTTP transfers. `CURLOPT_TIMEOUT` might be though.

Perhaps it would be better if I send you the source code?

**Comment by dfandrich** [comment]
What resolver is your libcurl using?

**Comment by bagder** [comment]
The verbose output would be interesting to see. Does it get stuck resolving the name? Does no names/URLs work?

**Comment by Jacellen** [comment]
> The verbose output would be interesting to see. Does it get stuck resolving the name? Does no names/URLs work?

Yes, it gets stuck on the domain, everything is fine with the IP

**Comment by bagder** [comment]
If it gets stuck while name resolving `CURLOPT_CONNECTTIMEOUT` set to 5 should make it cancel after 5 seconds. There seems to be something truly wrong with your system if it does not. Also, the mere fact that you cannot resolve host names is an indication things are not "normal".

**Comment by Jacellen** [comment]
> If it gets stuck while name resolving `CURLOPT_CONNECTTIMEOUT` set to 5 should make it cancel after 5 seconds. There seems to be something truly wrong with your system if it does not. Also, the mere fact that you cannot resolve host names is an indication things are not "normal".

Strange, but with std::thread(downloadstr).detach(); everything works fine, but not in the main thread

**Comment by bagder** [comment]
There is nothing for us to act on here. We run lots of tests on Windows all the time and we never see this behavior. Not to mention that quite a few users use it fine on Windows all the time. We can't reproduce and there aren't details provided to a level that can give us clues to otherwise figure it out.

**Comment by bagder** [comment]
Are you doing this in a DLL? If so, see #12577

**Comment by bagder** [comment]
I don't see what we can do about this.

**Comment by jay** [comment]
The reporter hasn't given us enough info to determine if this is a bug or not. Closing.
